README.md
## 2key.network 

### 2key.network Smart-Contracts architecture

[Official Smart-Contracts documentation](https://2key.github.io/contracts) is generated automatically based on description annotation
above all methods & contract implementations by [Solidity-docgen](https://github.com/OpenZeppelin/solidity-docgen) 
and website generated based on the output of Solidity-docgen using [Docosaurus](https://github.com/facebook/Docusaurus)


All contracts can be found [here](contracts/2key)

For development instructions, and how to run contracts locally, check [Setup dev-environment instructions](readmes/SetupDevEnv.md)

**Important**

*Bear in mind that all addresses in .env and .env.2key files are just for testing purpose!! Don't ever send any ETH or ERC20 token, since their private keys and mnemonics are publicly exposed!*
 
![TwoKeyInfrastructure](SC-architecture.png)


CHANGELOG.md
<a name="unreleased"></a>
## [Unreleased]


<a name="v1.4.201-develop"></a>
## [v1.4.201-develop] - 2021-08-04

<a name="v1.4.200-develop"></a>
## [v1.4.200-develop] - 2021-08-04

<a name="v1.4.194-develop"></a>
## [v1.4.194-develop] - 2021-08-04

<a name="v1.4.193-develop"></a>
## [v1.4.193-develop] - 2021-08-04

<a name="v1.4.192-develop"></a>
## [v1.4.192-develop] - 2021-08-03

<a name="v1.4.40-master"></a>
## [v1.4.40-master] - 2021-03-09

<a name="v1.4.39-master"></a>
## [v1.4.39-master] - 2021-03-04

<a name="v1.4.38-master"></a>
## [v1.4.38-master] - 2021-03-03

<a name="v1.4.37-master"></a>
## [v1.4.37-master] - 2021-02-27

<a name="v1.4.191-develop"></a>
## [v1.4.191-develop] - 2021-02-16

<a name="v1.2.79-staging"></a>
## [v1.2.79-staging] - 2021-02-15

<a name="v1.2.78-staging"></a>
## [v1.2.78-staging] - 2021-02-15

<a name="v1.4.190-develop"></a>
## [v1.4.190-develop] - 2021-01-20

<a name="v1.4.189-develop"></a>
## [v1.4.189-develop] - 2021-01-20
### Reverts
- Contracts deployed to public.staging-hdwallet Jan 20, 2021 2:48 PM


<a name="v1.4.36-master"></a>
## [v1.4.36-master] - 2021-01-19

<a name="v1.2.77-staging"></a>
## [v1.2.77-staging] - 2021-01-15

<a name="v1.4.35-master"></a>
## [v1.4.35-master] - 2021-01-14

<a name="v1.4.34-master"></a>
## [v1.4.34-master] - 2021-01-14

<a name="v1.4.188-develop"></a>
## [v1.4.188-develop] - 2021-01-13

<a name="v1.2.76-staging"></a>
## [v1.2.76-staging] - 2021-01-13

<a name="v1.2.75-staging"></a>
## [v1.2.75-staging] - 2021-01-12

<a name="v1.2.74-staging"></a>
## [v1.2.74-staging] - 2021-01-12

<a name="v1.4.187-develop"></a>
## [v1.4.187-develop] - 2021-01-12

<a name="v1.4.186-develop"></a>
## [v1.4.186-develop] - 2021-01-12

<a name="v1.4.185-develop"></a>
## [v1.4.185-develop] - 2021-01-12

<a name="v1.4.184-develop"></a>
## [v1.4.184-develop] - 2021-01-12
### Pull Requests
- Merge pull request [#60](https://github.com/2key/contracts/issues/60) from 2key/oracles


<a name="v1.2.73-staging"></a>
## [v1.2.73-staging] - 2020-12-08

<a name="v1.4.33-master"></a>
## [v1.4.33-master] - 2020-12-06

<a name="v1.2.72-staging"></a>
## [v1.2.72-staging] - 2020-12-06

<a name="v1.4.32-master"></a>
## [v1.4.32-master] - 2020-12-05

<a name="v1.4.31-master"></a>
## [v1.4.31-master] - 2020-12-05

<a name="v1.4.30-master"></a>
## [v1.4.30-master] - 2020-12-05

<a name="v1.2.71-staging"></a>
## [v1.2.71-staging] - 2020-12-02

<a name="v1.2.70-staging"></a>
## [v1.2.70-staging] - 2020-12-01

<a name="v1.4.183-develop"></a>
## [v1.4.183-develop] - 2020-11-30

<a name="v1.2.69-staging"></a>
## [v1.2.69-staging] - 2020-11-30

<a name="v1.2.68-staging"></a>
## [v1.2.68-staging] - 2020-11-25

<a name="v1.4.182-develop"></a>
## [v1.4.182-develop] - 2020-11-25

<a name="v1.2.67-staging"></a>
## [v1.2.67-staging] - 2020-11-24

<a name="v1.4.181-develop"></a>
## [v1.4.181-develop] - 2020-11-24

<a name="v1.2.66-staging"></a>
## [v1.2.66-staging] - 2020-11-24

<a name="v1.4.180-develop"></a>
## [v1.4.180-develop] - 2020-11-23

<a name="v1.4.179-develop"></a>
## [v1.4.179-develop] - 2020-11-23

<a name="v1.4.178-develop"></a>
## [v1.4.178-develop] - 2020-11-21

<a name="v1.2.65-staging"></a>
## [v1.2.65-staging] - 2020-11-20

<a name="v1.4.29-master"></a>
## [v1.4.29-master] - 2020-11-19

<a name="v1.4.177-develop"></a>
## [v1.4.177-develop] - 2020-11-19

<a name="v1.4.176-develop"></a>
## [v1.4.176-develop] - 2020-11-18

<a name="v1.4.175-develop"></a>
## [v1.4.175-develop] - 2020-11-18

<a name="v1.4.174-develop"></a>
## [v1.4.174-develop] - 2020-11-18

<a name="v1.4.28-master"></a>
## [v1.4.28-master] - 2020-11-17

<a name="v1.4.173-develop"></a>
## [v1.4.173-develop] - 2020-11-17

<a name="v1.4.172-develop"></a>
## [v1.4.172-develop] - 2020-11-08

<a name="v1.4.171-develop"></a>
## [v1.4.171-develop] - 2020-11-08

<a name="v1.4.170-develop"></a>
## [v1.4.170-develop] - 2020-11-08

<a name="v1.4.169-develop"></a>
## [v1.4.169-develop] - 2020-11-04

<a name="v1.4.168-develop"></a>
## [v1.4.168-develop] - 2020-11-03

<a name="v1.2.64-staging"></a>
## [v1.2.64-staging] - 2020-10-31

<a name="v1.4.167-develop"></a>
## [v1.4.167-develop] - 2020-10-31

<a name="v1.4.27-master"></a>
## [v1.4.27-master] - 2020-10-29

<a name="v1.4.26-master"></a>
## [v1.4.26-master] - 2020-10-29

<a name="v1.4.166-develop"></a>
## [v1.4.166-develop] - 2020-10-28

<a name="v1.4.25-master"></a>
## [v1.4.25-master] - 2020-10-28

<a name="v1.2.63-staging"></a>
## [v1.2.63-staging] - 2020-10-27

<a name="v1.2.62-staging"></a>
## [v1.2.62-staging] - 2020-10-27

<a name="v1.4.165-develop"></a>
## [v1.4.165-develop] - 2020-10-27

<a name="v1.4.24-master"></a>
## [v1.4.24-master] - 2020-10-26

<a name="v1.4.23-master"></a>
## [v1.4.23-master] - 2020-10-25

<a name="v1.4.22-master"></a>
## [v1.4.22-master] - 2020-10-25

<a name="v1.4.21-master"></a>
## [v1.4.21-master] - 2020-10-25

<a name="v1.4.164-develop"></a>
## [v1.4.164-develop] - 2020-10-21

<a name="v1.4.163-develop"></a>
## [v1.4.163-develop] - 2020-10-21

<a name="v1.2.61-staging"></a>
## [v1.2.61-staging] - 2020-10-15

<a name="v1.4.162-develop"></a>
## [v1.4.162-develop] - 2020-10-15

<a name="v1.4.161-develop"></a>
## [v1.4.161-develop] - 2020-10-14

<a name="v1.2.60-staging"></a>
## [v1.2.60-staging] - 2020-10-14

<a name="v1.4.160-develop"></a>
## [v1.4.160-develop] - 2020-10-13

<a name="v1.2.59-staging"></a>
## [v1.2.59-staging] - 2020-10-12

<a name="v1.4.159-develop"></a>
## [v1.4.159-develop] - 2020-10-12

<a name="v1.2.58-staging"></a>
## [v1.2.58-staging] - 2020-10-08

<a name="v1.2.57-staging"></a>
## [v1.2.57-staging] - 2020-10-08

<a name="v1.2.56-staging"></a>
## [v1.2.56-staging] - 2020-10-08

<a name="v1.4.158-develop"></a>
## [v1.4.158-develop] - 2020-10-05

<a name="v1.4.157-develop"></a>
## [v1.4.157-develop] - 2020-10-01

<a name="v1.4.156-develop"></a>
## [v1.4.156-develop] - 2020-10-01

<a name="v1.4.155-develop"></a>
## [v1.4.155-develop] - 2020-09-30

<a name="v1.4.154-develop"></a>

helpers.js
const childProcess = require('child_process');
const rimraf = require('rimraf');
const simpleGit = require('simple-git/promise');
const path = require('path');
const fs = require('fs');
const axios = require('axios');
const IPFS = require('ipfs-http-client');
const ipfs = new IPFS('ipfs.2key.net', 443, { protocol: 'https' });

require('dotenv').config({ path: path.resolve(process.cwd(), './.env-slack')});

const env_to_channelCode = {
    "test": "CKL4T7M2S",
    "staging": "CKKRPNR55",
    "prod": "CKHG3LS20"
};

const branch_to_env = {
    "develop": "test",
    "staging": "staging",
    "master": "prod"
};

const incrementVersion = ((version) => {
    if(version == "") {
        version = "1.0.0";
        return version;
    }
    let vParts = version.split('.');
    if(vParts.length < 2) {
        vParts = "1.0.0".split('.');
    }
    // assign each substring a position within our array
    let partsArray = {
        major : vParts[0],
        minor : vParts[1],
        patch : vParts[2]
    };
    // target the substring we want to increment on
    partsArray.patch = parseFloat(partsArray.patch) + 1;
    // set an empty array to join our substring values back to
    let vArray = [];
    // grabs each property inside our partsArray object
    for (let prop in partsArray) {
        if (partsArray.hasOwnProperty(prop)) {
            // add each property to the end of our new array
            vArray.push(partsArray[prop]);
        }
    }
    // join everything back into one string with a period between each new property
    let newVersion = vArray.join('.');
    return newVersion;
});

const runProcess = (app, args) => new Promise((resolve, reject) => {
    console.log('Run process', app, args && args.join(' '));
    const proc = childProcess.spawn(app, args, {stdio: [process.stdin, process.stdout, process.stderr]});
    proc.on('close', async (code) => {
        console.log('process exit with code', code);
        if (code === 0) {
            resolve(code);
        } else {
            reject(code);
        }
    });
});

/**
 * This is function to run when we want to update our campaigns
 * @param network
 * @returns {Promise<any>}
 */
const runDeployTokenSellCampaignMigration = (network) => new Promise(async(resolve, reject) => {
    try {
        await runProcess(path.join(__dirname, 'node_modules/.bin/truffle'), ['migrate', '--f', '6', '--to', '6', '--network', network]);
        resolve(true);
    } catch (e) {
        reject(e);
    }
});


const runDeployDonationCampaignMigration = (network) => new Promise(async(resolve, reject) => {
    try {
        await runProcess(path.join(__dirname, 'node_modules/.bin/truffle'), ['migrate', '--f', '7', '--to', '7', '--network', network]);
        resolve(true);
    } catch (e) {
        reject(e);
    }
});

/**
 * This is function to run when we want to update our cpc campaigns
 * @param network
 * @returns {Promise<any>}
 */
const runDeployCPCCampaignMigration = (network) => new Promise(async(resolve, reject) => {
    try {
        await runProcess(path.join(__dirname, 'node_modules/.bin/truffle'), ['migrate', '--f', '8', '--to', '8', '--network', network, 'update_cpc']);
        resolve(true);
    } catch (e) {
        reject(e);
    }
});

const runDeployCPCNoRewardsMigration = (network) => new Promise(async(resolve, reject) => {
    try {
        await runProcess(path.join(__dirname, 'node_modules/.bin/truffle'), ['migrate', '--f', '14', '--to', '14', '--network', network, 'update_cpc']);
        resolve(true);
    } catch (e) {
        reject(e);
    }
});


const runDeployPaymentHandlersMigration = (network) => new Promise(async (resolve, reject) => {
    try {
        await runProcess(path.join(__dirname, 'node_modules/.bin/truffle'), ['migrate', '--f', '15', '--to', '15', '--network', network, 'update_cpc']);
        resolve(true);
    } catch (e) {
        reject(e);
    }
});


const runDeployPlasmaParticipationsMining = (network) => new Promise(async (resolve, reject) => {
    try {
        await runProcess(path.join(__dirname, 'node_modules/.bin/truffle'), ['migrate', '--f', '16', '--to', '16', '--network', network]);
        resolve(true);
    } catch (e) {
        reject(e);
    }
});


/**
 * This is function to run when we want to update our campaigns
 * @param network
 * @returns {Promise<any>}
 */
const runDeployCPCFirstTime = (network) => new Promise(async(resolve, reject) => {
    try {
        await runProcess(path.join(__dirname, 'node_modules/.bin/truffle'), ['migrate', '--f', '9', '--to', '9', '--network', network]);
        resolve(true);
    } catch (e) {
        reject(e);
    }
});


/**
 * If there's a need to update, we'll run this function
 * @param network
 * @param contractName
 * @returns {Promise<any>}
 */
const runUpdateMigration = (network, contractName) => new Promise(async(resolve,reject) => {
    try {
        console.log("Running update migration");
        await runProcess(path.join(__dirname, 'node_modules/.bin/truffle'), ['migrate', '--f', '10', '--to', '10', '--network', network, 'update', contractName]);
        resolve(true);
    } catch (e) {
        reject(e);
    }
});


/**
 *
 * @param network
 * @param contractName
 * @returns {Promise<any>}
 */
const runDeployPlasmaEventSourceMigration = (network) => new Promise(async(resolve, reject) => {
    try {
        console.log("Running update migration");
        await runProcess(path.join(__dirname, 'node_modules/.bin/truffle'), ['migrate', '--f', '11', '--to', '11', '--network', network]);
        resolve(true);
    } catch (e) {
        reject(e);
    }
});



const runDeployFeeManagerMigration = (network) => new Promise(async(resolve, reject) => {
    try {
        await runProcess(path.join(__dirname, 'node_modules/.bin/truffle'), ['migrate', '--f', '12', '--to', '12', '--network', network]);
        resolve(true);
    } catch (e) {
        reject(e);
    }
});

const runDeployPlasmaReputation = (network) => new Promise(async(resolve, reject) => {
    try {
        await runProcess(path.join(__dirname, 'node_modules/.bin/truffle'), ['migrate', '--f', '13', '--to', '13', '--network', network]);
        resolve(true);
    } catch (e) {
        reject(e);
    }
});

const runDeployPPCNoRewards = (network) => new Promise(async(resolve, reject) => {
    try {
        await runProcess(path.join(__dirname, 'node_modules/.bin/truffle'), ['migrate', '--f', '14', '--to', '14', '--network', network]);
        resolve(true);
    } catch (e) {
        reject(e);
    }
});


/**
 * This is function to run when we want to update our campaigns
 * @param network
 * @returns {Promise<any>}
 */
const runTruffleCompile = () => new Promise(async(resolve, reject) => {
    try {
        await runProcess(path.join(__dirname, 'node_modules/.bin/truffle'), ['compile']);
        resolve(true);
    } catch (e) {
        reject(e);
    }
});



const checkArgumentsForUpdate = ((argumentsPassed) => {
    argumentsPassed.forEach((argument) => {
        if(argument == 'update') {
            return true;
        }
    });
    return false;
});

const checkIsHardRedeploy = ((argumentsPassed) => {
    argumentsPassed.forEach((argument) => {
        if(argument.toString() == '--reset') {s
            return true;
        }
    });
    return false;
});

/**
 *
 * @returns {Promise<any>}
 */
const getConfigForTheBranch = () => new Promise(async(resolve,reject) => {
    try {
        let branch = await getGitBranch();
        let filePath = `./2key-protocol/dist/contracts_deployed-${branch}.json`;
        let file = fs.readFileSync(filePath);
        resolve(JSON.parse(file));
    } catch (e) {
        reject(e);
    }
});

/**
 * Remove direcotry
 * @param dir
 * @returns {Promise<any>}
 */
const rmDir = (dir) => new Promise((resolve) => {
    rimraf(dir, () => {
        resolve();
    })
});

/**
 *
 * @returns {Promise<any>}
 */
const getGitBranch = () => new Promise(async(resolve,reject) => {
    try {
        const currentRepo = simpleGit();
        let branchStatus = await currentRepo.status();
        resolve(branchStatus.current);
    } catch (e) {
        reject(e);
    }
});


const slack_message_proposal_created = async (contractName, newVersion, proposalBytecode, network) => {

    const token = process.env.SLACK_TOKEN;

    const branch = await getGitBranch();
    const devEnv = branch_to_env[branch];

    const body = {
        channel: env_to_channelCode[devEnv],
        attachments: [
            {
                blocks: [
                    {
                        type: 'section',
                        text: {
                            type: 'mrkdwn',
                            text: `*Deployed new version of*  \`${contractName}\` *to network:* ${network} --> <@eiTan> <@Kiki> \n *New version :* \`${newVersion}\``
                        },

                    },
                    {
                        type: 'section',
                        text: {
                            type: 'mrkdwn',
                            text: proposalBytecode
                        },

                    },
                ],
            },
        ],

    };

    return axios.post('https://slack.com/api/chat.postMessage?parse=full&link_names=1', body, {
        headers: {
            'Authorization': `Bearer ${token}`,
            'Content-type': 'application/json; charset=utf-8'
        }
    });
};

/**
 * Function to send message to slack channel
 * @param message
 */
const slack_message = async (newVersion, oldVersion, devEnv) => {
    const token = process.env.SLACK_TOKEN;

    let commitHash = require('child_process')
        .execSync('git rev-parse HEAD')
        .toString().trim();


    let commitHash2keyProtocol = require('child_process')
        .execSync('cd 2key-protocol/dist && git rev-parse HEAD')
        .toString().trim();

    let diffData = require('child_process')
        .execSync(`git --no-pager log --no-color -p -1 ${oldVersion}..${newVersion} .`)
        .toString().trim();

    const body = {
        channel: env_to_channelCode[devEnv],
        attachments: [
            {
                blocks: [
                    {
                        type: 'section',
                        text: {
                            type: 'mrkdwn',
                            text: `[${devEnv.toUpperCase()}] protocol updated, version: ${newVersion}`,
                        },
                    },
                    {
                        type: 'section',
                        text: {
                            type: 'mrkdwn',
                            text: `\`\`\`${diffData}\`\`\``,
                        },
                    },
                    {
                        type: 'divider',
                    },
                    {
                        type: 'context',
                        elements: [
                            {
                                type: 'mrkdwn',
                                text: `For more info, click <https://github.com/2key/contracts/compare/${oldVersion}...${newVersion}|here>`,
                            },
                        ],
                    },
                    {
                        type: 'section',
                        text: {
                            type: 'mrkdwn',
                            text: `*Last commit contracts repo: * <https://github.com/2key/contracts/commit/${commitHash}|${commitHash}>`,
                        },
                    },
                    {
                        type: 'section',
                        text: {
                            type: 'mrkdwn',
                            text: `*Last commit 2key-protocol repo: * <https://github.com/2key/2key-protocol/commit/${commitHash2keyProtocol.toUpperCase()}|${commitHash2keyProtocol.toUpperCase()}>`,
                        },
                    },
                ],
            },
        ],
    };


    return axios.post('https://slack.com/api/chat.postMessage?parse=full&link_names=1', body, {
        headers: {
            'Authorization': `Bearer ${token}`,
            'Content-type': 'application/json; charset=utf-8'
        }
    });
};

/**
 * Helper function to perform mechanism of sorting per vesion
 * @param versionA
 * @param versionB
 * @returns {number}
 */
const sortMechanism = (versionA,versionB) => {
    versionA = versionA.split('-')[0].split('.');
    versionB = versionB.split('-')[0].split('.');
    if (parseInt(versionA[0]) > parseInt(versionB[0]))
        return 1;
    if (parseInt(versionA[0]) < parseInt(versionB[0]))
        return -1;
    else {
        if(parseInt(versionA[1]) > parseInt(versionB[1]))
            return 1;
        if(parseInt(versionA[1]) < parseInt(versionB[1]))
            return -1;
        else {
            if(parseInt(versionA[2]) > parseInt(versionB[2]))
                return 1;
            if(parseInt(versionA[2]) < parseInt(versionB[2]))
                return -1;
            else return 0;
        }
    }
};



const ipfsGet = (hash) => new Promise((resolve, reject) => {
    ipfs.get(hash, (err, res) => {
        if (err) {
            reject(err);
        } else {
            resolve(res[0] && res[0].content.toString());
        }
    });
});

const ipfsAdd = (data, deployment) => new Promise((resolve, reject) => {
    ipfs.add(ipfs.types.Buffer.from(data), { pin: deployment }, (err, res) => {
        if (err) {
            reject(err);
        } else {
            resolve(res);
        }
    });
});





module.exports = {
    incrementVersion,
    runProcess,
    runDeployTokenSellCampaignMigration,
    runDeployDonationCampaignMigration,
    runUpdateMigration,
    rmDir,
    getGitBranch,
    getConfigForTheBranch,
    slack_message_proposal_created,
    slack_message,
    checkIsHardRedeploy,
    checkArgumentsForUpdate,
    sortMechanism,
    ipfsAdd,
    ipfsGet,
    runDeployPlasmaEventSourceMigration,
    runDeployCPCCampaignMigration,
    runDeployFeeManagerMigration,
    runDeployCPCFirstTime,
    runTruffleCompile,
    runDeployPlasmaReputation,
    runDeployPPCNoRewards,
    runDeployCPCNoRewardsMigration,
    runDeployPaymentHandlersMigration,
    runDeployPlasmaParticipationsMining
};
